styx-client/src/main/java/com/spotify/styx/client/StyxApolloClient.java

```diff
@@ -0,0 +1,319 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+
+package com.spotify.styx.client;
+
+import static com.spotify.styx.serialization.Json.OBJECT_MAPPER;
+import static com.spotify.styx.serialization.Json.serialize;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ArrayNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.ImmutableList;
+import com.spotify.apollo.Client;
+import com.spotify.apollo.Request;
+import com.spotify.apollo.Response;
+import com.spotify.futures.CompletableFutures;
+import com.spotify.styx.api.BackfillPayload;
+import com.spotify.styx.api.BackfillsPayload;
+import com.spotify.styx.api.ResourcesPayload;
+import com.spotify.styx.api.RunStateDataPayload;
+import com.spotify.styx.model.Backfill;
+import com.spotify.styx.model.BackfillInput;
+import com.spotify.styx.model.Event;
+import com.spotify.styx.model.Resource;
+import com.spotify.styx.model.Workflow;
+import com.spotify.styx.model.WorkflowState;
+import com.spotify.styx.model.data.EventInfo;
+import com.spotify.styx.util.EventUtil;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletionStage;
+import java.util.stream.Collectors;
+import okio.ByteString;
+
+/**
+ * Styx Apollo Client Implementation
+ */
+public class StyxApolloClient
+    implements StyxWorkflowClient, StyxBackfillClient, StyxResourceClient,
+               StyxSchedulerClient, StyxStatusClient {
+  private static final String STYX_API_ENDPOINT = "/api/v2";
+  private static final String UTF_8 = "UTF-8";
+  private static final int TTL_SECONDS = 90;
+
+  private final Client client;
+  private final String apiHost;
+  private final String userAgentIdentifier;
+
+  public StyxApolloClient(final Client client,
+                          final String apiHost,
+                          final String userAgentIdentifier) {
+    this.client = client;
+    this.apiHost = apiHost;
+    this.userAgentIdentifier = userAgentIdentifier;
+  }
+
+  @Override
+  public CompletionStage<RunStateDataPayload> activeStates(Optional<String> componentId) {
+    String url = apiUrl("status", "activeStates");
```

Comment:
This kind of URL handling feels quite primitive, we should probably switch to https://square.github.io/okhttp/3.x/okhttp/okhttp3/HttpUrl.html. Have a `TODO` maybe? Not to get scope creep.